neuro_san/service/usage/wrapped_usage_logger.py

```diff
@@ -0,0 +1,130 @@
+                                    can come from model costs from libraries instead of directly from
+                                    providers.
+                    "total_tokens" - Total tokens used for the request.
+
+                More keys can appear, but should not be counted on.
+                The ones listed above contain potentially salient information for usage logging purposes.
+
+        :param request_metadata: A dictionary of filtered request metadata whose keys contain
+                identifying information for the usage log.
+        """
+        if self.wrapped is None:
+            # Nothing to report
+            return
+
+        if token_dict is None:
+            # Nothing to report
+            return
+
+        compliant_token_dict: Dict[str, Any] = self.make_compliant_token_dict(token_dict)
+        minimal_metadata: Dict[str, Any] = self.minimize_metadata(request_metadata)
+
+        await self.wrapped.log_usage(compliant_token_dict, minimal_metadata)
+
+    def synchronous_log_usage(self, token_dict: Dict[str, Any], request_metadata: Dict[str, Any]):
+        """
+        Logs the token usage for external capture.
+        See comments for log_usage() above.
+        """
+        run(self.log_usage(token_dict, request_metadata))
+
+    def make_compliant_token_dict(self, token_dict: Dict[str, Any]) -> Dict[str, Any]:
+        """
+        :param token_dict: The token dictionary to make compliant if it is not already
+        :return: A token dictionary compliant to the UsageLogger interface
+        """
+        compliant: Dict[str, Any] = token_dict
+
+        if "total_tokens" in token_dict.keys():
+            # We have a raw token dictionary without any attribution to the LLM.
+            # Make the dictionary compliant
+            compliant = {
+                "all": token_dict
+            }
+
+        return compliant
+
+    def minimize_metadata(self, request_metadata: Dict[str, Any]) -> Dict[str, Any]:
+        """
+        :param request_metadata: The raw request metadata dictionary that could easily contain
+                    more keys than we want to send to the UsageLogger.
+        :return: A minimized dictionary that only sends the keys we need to the UsageLogger.
+                The idea is that this prevents the UsageLogger from getting potentially
+                sensitive information it shouldn't really have.
+
+                If the requested keys in the metadata are not there, they will also not appear
+                in the returned minimized dictionary.
+        """
+        minimized: Dict[str, Any] = {}
+
+        # Try getting the value from the more specific env var before falling back to the
+        # other env var.
+        keys_string: str = environ.get("AGENT_USAGE_LOGGER_METADATA",
+                                       environ.get("AGENT_FORWARDED_REQUEST_METADATA"))
```

Comment:
Check the environment variables to make a need-to-know version of the request metadata dictionary that comes along with every request.